csync/tests/vio_tests/check_vio_ext.c

```diff
@@ -0,0 +1,335 @@
+
+      assert_int_equal(rc, 0);
+      p[i] = '/';
+    }
+    i++;
+  }
+  SAFE_FREE(mypath);
+}
+
+/*
+ * This function uses the vio_opendir, vio_readdir and vio_closedir functions
+ * to traverse a file tree that was created before by the create_dir function.
+ *
+ * It appends a listing to the result member of the incoming struct in *state
+ * that can be compared later to what was expected in the calling functions.
+ *
+ */
+static void traverse_dir(void **state, const char *dir)
+{
+    csync_vio_handle_t *dh;
+    csync_vio_file_stat_t *dirent;
+    statevar *sv = (statevar*) *state;
+    CSYNC *csync = sv->csync;
+    char *subdir;
+    char *subdir_out;
+    int rc;
+    int is_dir;
+
+    /* Format: Smuggle in the C: for unix platforms as its urgently needed
+     * on Windows and the test can be nicely cross platform this way. */
+#ifdef _WIN32
+    const char *format_str = "%s %s";
+#else
+    const char *format_str = "%s C:%s";
+#endif
+
+    dh = csync_vio_opendir(csync, dir);
+    assert_non_null(dh);
+
+    while( (dirent = csync_vio_readdir(csync, dh)) ) {
+        assert_non_null(dirent->name);
+
+        assert_int_equal( dirent->fields & CSYNC_VIO_FILE_STAT_FIELDS_TYPE, CSYNC_VIO_FILE_STAT_FIELDS_TYPE );
+
+        if( c_streq( dirent->name, "..") || c_streq( dirent->name, "." )) {
+          continue;
+        }
+
+        is_dir = (dirent->type == CSYNC_VIO_FILE_TYPE_DIRECTORY) ? 1:0;
+
+        assert_int_not_equal( asprintf( &subdir, "%s/%s", dir, dirent->name ), -1 );
+
+        assert_int_not_equal( asprintf( &subdir_out, format_str,
+                                        is_dir ? "<DIR>":"    ",
+                                        subdir), -1 );
+
+        if( !sv->result ) {
+            sv->result = c_strdup( subdir_out);
+        } else {
+            int newlen = 1+strlen(sv->result)+strlen(subdir_out);
+            char *tmp = sv->result;
+            sv->result = c_malloc(newlen);
+            strcpy( sv->result, tmp);
+            SAFE_FREE(tmp);
+
+            strcat( sv->result, subdir_out );
+        }
+        printf("%s\n", subdir_out);
+        traverse_dir( state, subdir);
+
+        SAFE_FREE(subdir);
+        SAFE_FREE(subdir_out);
+    }
+
+    csync_vio_file_stat_destroy(dirent);
+    rc = csync_vio_closedir(csync, dh);
+    assert_int_equal(rc, 0);
+
+}
+
+static void check_readdir_shorttree(void **state)
+{
+    statevar *sv = (statevar*) *state;
+
+    const char *t1 = "alibaba/und/die/vierzig/räuber/";
+    create_dirs( t1 );
+
+    traverse_dir(state, CSYNC_TEST_DIR);
+
+    assert_string_equal( sv->result,
+                         "<DIR> C:/tmp/csync_test/alibaba"
+                         "<DIR> C:/tmp/csync_test/alibaba/und"
+                         "<DIR> C:/tmp/csync_test/alibaba/und/die"
+                         "<DIR> C:/tmp/csync_test/alibaba/und/die/vierzig"
+                         "<DIR> C:/tmp/csync_test/alibaba/und/die/vierzig/räuber" );
```

Comment:
Could you also test special characters with more width? Like Chinese or Cyrillic Characters?